### PR TITLE
Ps 5.7 4196

### PR DESCRIPTION
**Pull request by george-lorch**
Prepare MyRocks for 8.0

### CODE REVIEW

**Comment by george-lorch** [comment]
https://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1860/

File: storage/rocksdb/ha_rocksdb.cc [hunk 1]
```diff
@@ -4455,44 +4416,43 @@ static inline bool is_valid(rocksdb::Iterator *scan_it) {
...
-  length = (uint)strlen(table_name);
+
+  const std::string s_table_name(table_name);
 
   // First, look up the table in the hash map.
   RDB_MUTEX_LOCK_CHECK(m_mutex);
-  if (!(table_handler = reinterpret_cast<Rdb_table_handler *>(my_hash_search(
-            &m_hash, reinterpret_cast<const uchar *>(table_name), length)))) {
+  Rdb_table_handler *table_handler = nullptr;
+  auto it = m_hash.find(s_table_name);
```
Comment (laurynas-biveinis) [inline storage/rocksdb/ha_rocksdb.cc]: const auto & it

File: storage/rocksdb/ha_rocksdb.cc [hunk 2]
```diff
@@ -4793,10 +4750,9 @@ void Rdb_open_tables_map::release_table_handler(
   DBUG_ASSERT(table_handler != nullptr);
   DBUG_ASSERT(table_handler->m_ref_count > 0);
   if (!--table_handler->m_ref_count) {
-    // Last rereference was released. Tear down the hash entry.
-    const auto ret MY_ATTRIBUTE((__unused__)) =
-        my_hash_delete(&m_hash, reinterpret_cast<uchar *>(table_handler));
-    DBUG_ASSERT(!ret); // the hash entry must actually be found and deleted
+    const size_t ret MY_ATTRIBUTE((__unused__)) =
```
Comment (laurynas-biveinis) [inline storage/rocksdb/ha_rocksdb.cc]: No need to change from auto to size_t

File: storage/rocksdb/ha_rocksdb.cc [hunk 3]
```diff
@@ -4510,18 +4470,15 @@ std::vector<std::string> rdb_get_open_table_names(void) {
...
 std::vector<std::string> Rdb_open_tables_map::get_table_names(void) const {
-  ulong i;
   const Rdb_table_handler *table_handler;
   std::vector<std::string> names;
 
   RDB_MUTEX_LOCK_CHECK(m_mutex);
-  for (i = 0; (table_handler = reinterpret_cast<const Rdb_table_handler *>(
-                   my_hash_const_element(&m_hash, i)));
-       i++) {
+  for (auto it = m_hash.begin(); it != m_hash.end(); it++) {
```
Comment (laurynas-biveinis) [inline storage/rocksdb/ha_rocksdb.cc]: const auto & it : m_hash (or if you don't like range loop, s/begin/cbegin, s/end/cend

File: storage/rocksdb/rdb_datadic.cc [hunk 4]
```diff
@@ -4027,19 +3998,21 @@ bool Rdb_ddl_manager::init(Rdb_dict_manager *const dict_arg,
...
+  Rdb_tbl_def *ret = nullptr;
+
   if (lock) {
     mysql_rwlock_rdlock(&m_rwlock);
   }
 
-  Rdb_tbl_def *const rec = reinterpret_cast<Rdb_tbl_def *>(my_hash_search(
-      &m_ddl_hash, reinterpret_cast<const uchar *>(table_name.c_str()),
-      table_name.size()));
+  auto it = m_ddl_hash.find(table_name);
```
Comment (laurynas-biveinis) [inline storage/rocksdb/rdb_datadic.cc]: const auto & it

File: storage/rocksdb/rdb_datadic.cc [hunk 5]
```diff
@@ -4204,21 +4177,19 @@ int Rdb_ddl_manager::put_and_write(Rdb_tbl_def *const tbl,
...
   if (lock)
     mysql_rwlock_wrlock(&m_rwlock);
 
   // We have to do this find because 'tbl' is not yet in the list.  We need
   // to find the one we are replacing ('rec')
-  rec = find(dbname_tablename, false);
-  if (rec) {
-    // this will free the old record.
-    my_hash_delete(&m_ddl_hash, reinterpret_cast<uchar *>(rec));
+  auto it = m_ddl_hash.find(dbname_tablename);
```
Comment (laurynas-biveinis) [inline storage/rocksdb/rdb_datadic.cc]: const auto & it

File: storage/rocksdb/rdb_datadic.cc [hunk 6]
```diff
@@ -4249,8 +4220,11 @@ void Rdb_ddl_manager::remove(Rdb_tbl_def *const tbl,
   const rocksdb::Slice tkey((char *)buf, pos);
   m_dict->delete_key(batch, tkey);
 
-  /* The following will also delete the object: */
-  my_hash_delete(&m_ddl_hash, reinterpret_cast<uchar *>(tbl));
+  auto it = m_ddl_hash.find(dbname_tablename);
```
Comment (laurynas-biveinis) [inline storage/rocksdb/rdb_datadic.cc]: const auto & it

File: storage/rocksdb/rdb_datadic.cc [hunk 7]
```diff
@@ -4303,14 +4277,17 @@ bool Rdb_ddl_manager::rename(const std::string &from, const std::string &to,
 }
 
 void Rdb_ddl_manager::cleanup() {
-  my_hash_free(&m_ddl_hash);
+  for (auto it = m_ddl_hash.cbegin(); it != m_ddl_hash.cend(); it++)
```
Comment (laurynas-biveinis) [inline storage/rocksdb/rdb_datadic.cc]: const auto & it, also consider switching to range-based loop const auto & it : m_ddl_hash

File: storage/rocksdb/rdb_datadic.cc [hunk 8]
```diff
@@ -4319,9 +4296,8 @@ int Rdb_ddl_manager::scan_for_tables(Rdb_tables_scanner *const tables_scanner) {
   ret = 0;
   i = 0;
 
-  while ((
-      rec = reinterpret_cast<Rdb_tbl_def *>(my_hash_element(&m_ddl_hash, i)))) {
-    ret = tables_scanner->add_table(rec);
+  for (auto it = m_ddl_hash.cbegin(); it != m_ddl_hash.cend(); it++) {
```
Comment (laurynas-biveinis) [inline storage/rocksdb/rdb_datadic.cc]: const auto & it, also consider switching to range-based loop

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]
The majority of comments ask to convert auto iterator ... to const auto & iterator ... - to avoid copying the iterator

**Comment by george-lorch** [comment]
All good points and addressed, new jenkins here https://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1861/

**Review by laurynas-biveinis — APPROVED** [review]

**Comment by george-lorch** [comment]
The previous Travis run identified a few minor formatting issues that I fixed, tested, pushed, jenkins https://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1864/ but not all are corrected as I disagree with the suggested reformat. Merging now.

**Pull request merged**